Optimize control code

JoergAtGithub opened the pull request:
This PR collects several low level optimizations for our control code. See the single commit messages.

We have currently 36816 ControlDoublePrivate instances in Mixxx the memory realignment reduces the size from 464 byte to 448 byte, this translates to 0.5 MegaByte heap memory reduction.
Furthermore the first fields are aligned to fit into a 128 byte SSE register. But I guess there are very few operation, which can benefit from an SSE copy etc. (in theory we could further optimize for SSE, by reducing ControlValueAtomic<double> from 136byte to 128byte).

Old memory alignment:
![ControlDoublePrivateBefore](https://github.com/mixxxdj/mixxx/assets/64457745/581ae4e0-c8d0-471f-b33d-96941cffe79b)


New memory alignment:
![ControlDoublePrivateThisPR](https://github.com/mixxxdj/mixxx/assets/64457745/febceb7e-c0ca-4f17-82f4-05c6fb6d387c)

JoergAtGithub committed:
Reserve memory m_controls before filling it in loop

JoergAtGithub committed:
Moved duplicated code in new private function setBehavior

JoergAtGithub committed:
Memory alignment of ControlDoublePrivate to reduce size from 464 byte to 448 byte
First fields fit in 128 byte SSE register now

daschuer commented:
The reason that ControlDoublePrivate is so big is a nasty regression. It should be only for bites. The fix is here:  https://github.com/mixxxdj/mixxx/pull/13355

daschuer commented on src/control/control.h:
```diff
@@ -183,8 +183,29 @@ class ControlDoublePrivate : public QObject {
+    // The control value.
+    ControlValueAtomic<double> m_value;
+    // The default control value.
+    ControlValueAtomic<double> m_defaultValue;
```
How do we protect that for get cluttered again? Maybe define regions via comments?

Swiftb0y replied:
We could try `-Wpadded` https://developers.redhat.com/blog/2016/06/01/how-to-avoid-wasting-megabytes-of-memory-a-few-bytes-at-a-time#tools

JoergAtGithub replied:
I tried -Wpadded on CI, but it does not only warn about padding inbetween the data structure, but also if the overall size is not matching the byte alignment size, e.g. any class with 6 bytes would need to filled up with 2 explicit padding bytes.

I found also a currently unused macro in https://github.com/mixxxdj/mixxx/blob/15b4444584b9cd2514946d6a645746cae3f5a315/src/util/platform.h#L15

Maybe someone on Linux could try: https://linux.die.net/man/1/pahole

Swiftb0y replied:
Okay that makes sense I guess when the compiler expects you to nest these packed data structures. The only other potential solution I found was to generate two versions of the type, one packed (potentially with suboptimal alignment and then `static_assert(sizeof(T) == sizeof(_T_packed)) (potentially with a bit of rounding to manually account for the overall size). I wouldn't know how to accomplish that without copy pasting or a macro (doing the copy-pasting automatically) though.

daschuer replied:
Let's not over engineer it. Just a comment state informs there reader that the oder of the variables has been chosen to avoid memory gaps due to alignment is probably sufficient.

JoergAtGithub replied:
It's already added: https://github.com/mixxxdj/mixxx/blob/6728c9156caaa904698da5cef3f723279a5129ed/src/control/control.h#L203-L205

daschuer commented on src/control/controlobject.cpp:
```diff
@@ -14,7 +14,7 @@ ControlObject::ControlObject(const ConfigKey& key,
         bool bPersist,
         double defaultValue)
-        : m_key(key) {
+        : m_key(std::move(key)) {
```
We cannot move form a const &. I guess this std::move which boils down to a cast is ignored. 
In this case, we need to pass the key by value to make use of the full benefit of avoiding reference counting.

daschuer commented on src/control/controlobject.cpp:
```diff
@@ -103,7 +103,7 @@ void ControlObject::setParameterFrom(double v, QObject* pSender) {
 // static
 void ControlObject::set(const ConfigKey& key, const double& value) {
-    QSharedPointer<ControlDoublePrivate> pCop = ControlDoublePrivate::getControl(key);
+    auto pCop = ControlDoublePrivate::getControl(key);
```
The function name getControl() does not reveal that it returns a `QSharedPointer<ControlDoublePrivate>`  so we need to keep the full type of rename the function.

daschuer commented on src/control/controlpushbutton.cpp:
```diff
@@ -10,12 +10,7 @@ ControlPushButton::ControlPushButton(const ConfigKey& key, bool bPersist, double
-                        static_cast<ControlPushButtonBehavior::ButtonMode>(m_buttonMode),
-                        m_iNoStates));
-    }
+    setBehavior();
```
A setter usually takes a parameter. Maybe `updateBehaviour()`?  

A setBehaviour() function taking num_states and mode would be also beneficial for places like here, where the first behaviour is allocated and destroyed right away: 
https://github.com/mixxxdj/mixxx/blob/8a00cc00e8e5fad1dfaf5d16c1a4bfcd98c60f3e/src/engine/controls/vinylcontrolcontrol.cpp#L42-L43

daschuer commented on src/control/pollingcontrolproxy.h:
```diff
@@ -19,7 +19,7 @@ class PollingControlProxy {
 
     PollingControlProxy(const ConfigKey& key, ControlFlags flags = ControlFlag::None) {
-        m_pControl = ControlDoublePrivate::getControl(key, flags);
+        m_pControl = ControlDoublePrivate::getControl(std::move(key), flags);
```
getControl does not always keep the key, so moving has no benefit here. Conditionally copy form a const & should work best.

JoergAtGithub commented:
I merged in the changes from #13355, but I noticed, that the memory consumption of ControlDoublePrivate is much bigger in this Main branch, than in the 2.4 branch.

2.4:
![grafik](https://github.com/mixxxdj/mixxx/assets/64457745/731ca236-c084-472b-be83-f5d1e388054d)

Main:
![grafik](https://github.com/mixxxdj/mixxx/assets/64457745/9d64f8c5-8315-495f-b52a-a045595c6b1f)

This PR:
![grafik](https://github.com/mixxxdj/mixxx/assets/64457745/3cdad28e-a825-47a1-af9c-baad9cdb2052)

daschuer commented:
I assume you build main with Qt6 and 2.4 with Qt5. Maybe the `sizeof(QString)` has changed due to the Small String Optimization?

JoergAtGithub commented:
> I assume you build main with Qt6 and 2.4 with Qt5. 

Yes, I did!

daschuer commented:
Qt6 has no SSO. The number in increased due to the use of a common QArrayDataPointer with holding two pointers one the header for implicit sharing, one to the actual data and one to the size which is now also 64 bit. This allows faster read only access and allows to the store strings longer than 2.147.483.648 characters (more than 480 Bibles)

daschuer commented on src/control/control.h:
```diff
@@ -183,8 +183,29 @@ class ControlDoublePrivate : public QObject {
     QAtomicPointer<ControlObject> m_pCreatorCO;
 
+    QString m_trackKey;
+
```
```suggestion

// Note: keep the order of the members below to not introduce gaps due to memory alignment in this often used class. 
```

JoergAtGithub committed:
Add comment about memory alignment

Co-authored-by: Daniel Schürmann <daschuer@mixxx.org>

JoergAtGithub committed:
Renamed setBehavior to updateBehavior and added a real setBehavior function and made use of it where applicable.
Unified the two ButtonMode enums to a common enum class.

JoergAtGithub commented:
@daschuer Friendly Ping

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/main' into optimize_control_code

Swiftb0y commented on src/control/controlpushbutton.h:
```diff
@@ -27,42 +28,36 @@
         return m_buttonMode;
     }
-    void setButtonMode(enum ButtonMode mode);
+    void setButtonMode(enum ControlButtonMode mode);
     void setStates(int num_states);
+    void setBehavior(enum ControlButtonMode mode, int num_states);
```
I'm pretty sure prefixing the type with `enum` is only really there for C-compatibility.
```suggestion
    void setButtonMode(ControlButtonMode mode);
    void setStates(int num_states);
    void setBehavior(ControlButtonMode mode, int num_states);
```

Swiftb0y commented on src/control/controlpushbutton.cpp:
```diff
@@ -8,39 +8,50 @@
         m_pControl->setBehavior(
                 new ControlPushButtonBehavior(
-                        static_cast<ControlPushButtonBehavior::ButtonMode>(m_buttonMode),
+                        static_cast<ControlButtonMode>(m_buttonMode),
```
unnecessary cast now that the enum type got unified.

Swiftb0y commented on src/control/controlbuttonmode.h:
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+enum class ControlButtonMode {
+    PUSH = 0,
+    TOGGLE,
+    POWERWINDOW,
+    LONGPRESSLATCHING,
+    TRIGGER
+};
```
since we're already touching every spot anyways. lets adhere to the core guidelines:
[Enum.5: Don’t use ALL_CAPS for enumerators](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-caps) and [Enum.8: Specify enumerator values only when necessary](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-value)
```suggestion
enum class ControlButtonMode {
    Push,
    Toggle,
    PowerWindow,
    LongPressLatching,
    Trigger
};
```

Swiftb0y commented on src/control/controlpushbutton.h:
```diff
@@ -27,42 +28,36 @@
-                return "TRIGGER";
-            default:
-                return "UNKNOWN";
+        case ControlButtonMode::PUSH:
+            return "PUSH";
+        case ControlButtonMode::TOGGLE:
+            return "TOGGLE";
+        case ControlButtonMode::POWERWINDOW:
+            return "POWERWINDOW";
+        case ControlButtonMode::LONGPRESSLATCHING:
+            return "LONGPRESSLATCHING";
+        case ControlButtonMode::TRIGGER:
+            return "TRIGGER";
+        default:
+            return "UNKNOWN";
```
do these strings require stability? Can we change them? lets remove the default case because it should not be possible to pass an invalid value and when the enum gets changed, we get notified thanks to `-Wswitch`. Also making `ControlButtonMode` a Q_ENUM would make this redundant, wdyt?
```suggestion
        case ControlButtonMode::Push:
            return QStringLiteral("Push");
        case ControlButtonMode::Toggle:
            return QStringLiteral("Toggle");
        case ControlButtonMode::PowerWindow:
            return QStringLiteral("PowerWindow");
        case ControlButtonMode::LongPressLatching:
            return QStringLiteral("LongPressLatching");
        case ControlButtonMode::Trigger:
            return QStringLiteral("Trigger");
```

JoergAtGithub replied:
Q_ENUM did not work here, because this is a static member function and Q_ENUM works only for membar variables.

Swiftb0y reviewed (commented):
a couple comments, thanks for the cleanup.

JoergAtGithub committed:
Removed unecessary keyword enum

JoergAtGithub committed:
Removed redundant cast

JoergAtGithub committed:
Renamed enum entries accrding cpp core guidelines

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/main' into optimize_control_code

JoergAtGithub committed:
Updated buttonModeToString to use QStringLiteral and added DEBUG_ASSERT for unexpected ControlButtonMode values.
Replaced header guard with #pragma once. Removed author and GPL
comment blocks.

Swiftb0y commented on src/control/controlpushbutton.h:
```diff
@@ -1,69 +1,42 @@
+            return QStringLiteral("LongPressLatching");
+        case ControlButtonMode::Trigger:
+            return QStringLiteral("Trigger");
+        default:
+            // Without the default branch we get compiler warnings
+            DEBUG_ASSERT(false);
+            return "Unknown";
         }
```
I assume you are referring to "control reaches end of non-void function [-Wreturn-type]"? If you just do this: we'll still get `-Wswitch` (which we want), but not `-Wreturn-type`.
```suggestion
        }
        DEBUG_ASSERT(false);
        return QStringLiteral("Unknown");
```

JoergAtGithub committed:
Moved handling of default path out of the switch statement

Swiftb0y reviewed (approved):
LGTM. Thank you.

Swiftb0y commented:
@daschuer please resolve your review

JoergAtGithub committed:
Use namespace for control button mode

daschuer commented:
Thank you!
